examples/vision/siamese.py

```diff
@@ -0,0 +1,351 @@
+# Euclidean distance = sqrt(sum(square(t1-t2)))
+def euclidean_distance(vects):
+    """
+    Find the Euclidean distance between two vectors
+
+    Parameter
+    ---------
+    vect  : list
+            list containing two tensors of same length
+
+    Return
+    ------
+    Tensor containing euclidean distance
+    (as floating point value) between vectors
+    """
+
+    x, y = vects
+    sum_square = tf.math.reduce_sum(tf.math.square(x - y), axis=1, keepdims=True)
+    return tf.math.sqrt(tf.math.maximum(sum_square, tf.keras.backend.epsilon()))
+
+
+
+input = layers.Input((28, 28, 1))
+x = tf.keras.layers.BatchNormalization()(input)
+x = layers.Conv2D(4, (5, 5), activation="tanh")(x)
+x = layers.AveragePooling2D(pool_size=(2, 2))(x)
+x = layers.Conv2D(16, (5, 5), activation="tanh")(x)
+x = layers.AveragePooling2D(pool_size=(2, 2))(x)
+x = layers.Flatten()(x)
+
+x = tf.keras.layers.BatchNormalization()(x)
+x = layers.Dense(10, activation="tanh")(x)
+embedding_network = keras.Model(input, x)
+
+
+input_1 = layers.Input((28, 28, 1))
+input_2 = layers.Input((28, 28, 1))
+
+# As mentioned above, Siamese Network share weights between
+# tower networks (sister networks). To allow this, we will use
+# same embedding network for both tower networks.
+tower_1 = embedding_network(input_1)
+tower_2 = embedding_network(input_2)
+
+merge_layer = layers.Lambda(euclidean_distance)([tower_1, tower_2])
+normal_layer = tf.keras.layers.BatchNormalization()(merge_layer)
+output_layer = layers.Dense(1, activation="sigmoid")(normal_layer)
+siamese = keras.Model(inputs=[input_1, input_2], outputs=output_layer)
+
+
+"""
+Define Constrastive Loss
+"""
+
+# Contrastive loss = mean( (1-true_value) * square(prediction) +
+#                         true_value * square( max(margin-prediction, 0) ))
+def contrastive_loss(y_true, y_pred, margin=1):
+    """
+    Calculate the constrastive loss
+
+    Parameters
+    ----------
+    margin  : int
+              margin defines the baseline for distance for which pairs
+              should be classified as dissimilar. - (default is 1)
+    y_true  : list
+              list of labels, each label is of type float32
+    y_pred  : list
+              list of predictions of same length as of y_true,
+              each label is of type float32
+
+    Return
+    ------
+    A tensor containing constrastive loss as floating point value
+    """
+
+    square_pred = tf.math.square(y_pred)
+    margin_square = tf.math.square(tf.math.maximum(margin - (y_pred), 0))
+    return tf.math.reduce_mean((1 - y_true) * square_pred + (y_true) * margin_square)
+
+
+
+"""
+Compile the model with constrastive loss
+"""
+
+siamese.compile(loss=contrastive_loss, optimizer="RMSprop", metrics=["accuracy"])
+
+siamese.summary()
+
+
+"""
+Train the model
+"""
+
+# Rarely it stucks at local optima, in that case just try again
+siamese.fit(
+    [x_train_1, x_train_2], labels_train, batch_size=16, validation_split=0.1, epochs=6
```

Comment:
Here it is wronw to use validation split. You must split it explicitely before hand to make sure train and validation splits are in fact independent. Otherwise, you are using the same images for validation and training. This is because of the nature of your sets: you build the pairs randomly.

The solution is to split the original training set into train and validation, and then for each of those, build the validation and training pairs.

Reply:
I think `validation_split` is ok, because even though i make pairs randomly but they are made before passing data to fit.
test of proof could be that `test_accuracy` is same as `valid_accuracy`, even better. 

if it were using same data then wouldn't it be overfitting on `valid_split` and giving `valid_accuracy` higher than `test_accuracy`?

Reply:
I didn't run them again to check. 

That said, the fact that metrics look ok doesn't mean they are ok. It may be due to the easiness of this particular problem that all metrics looks ok, but methodologically, training and validation sets are not independent, it's a fact.

In addition to this, since this is an example which is often read by non-experts, I think that taking care of those details is crucial.

It would be nice to hear a third opinion maybe.

Reply:
Done!